scripts/quality_test.py

```diff
@@ -190,81 +204,231 @@ def run_eslint(options):
+            for line in lines:
+                failure_match = failure_regex.match(line)
+                if failure_match:
+                    pii_check_passed = False
+                    break
+
+            # Each line in lines already contains a newline.
+            full_log = ''.join(lines)
+    else:
+        fail_quality('pii', f'FAILURE: Log file could not be found: {filename}')
+
+    return (uncovered_models, pii_check_passed, full_log)
+
+
+def run_pii_check():
+    """
+    Guarantee that all Django models are PII-annotated.
+    """
+    REPO_ROOT = repo_root()
+    REPORT_DIR = REPO_ROOT / 'reports'
+    pii_report_name = 'pii'
+    default_report_dir = (REPORT_DIR / pii_report_name)
+    report_dir = default_report_dir
+    output_file = os.path.join(report_dir, 'pii_check_{}.report')
+    env_report = []
+    pii_check_passed = True
+
+    for env_name, env_settings_file in (("CMS", "cms.envs.test"), ("LMS", "lms.envs.test")):
+        try:
+            print(f"Running {env_name} PII Annotation check and report")
+            print("-" * 45)
+
+            run_output_file = str(output_file).format(env_name.lower())
+            os.makedirs(report_dir, exist_ok=True)
+
+            # Prepare the environment for the command
+            env = {
+                **os.environ,  # Include the current environment variables
+                "DJANGO_SETTINGS_MODULE": env_settings_file  # Set DJANGO_SETTINGS_MODULE for each environment
+            }
+
+            command = [
+                "code_annotations",
+                "django_find_annotations",
+                "--config_file", ".pii_annotations.yml",
+                "--report_path", str(report_dir),
+                "--app_name", env_name.lower()
+            ]
+
+            # Run the command without shell=True
+            with open(run_output_file, 'w') as report_file:
+                subprocess.run(
+                    command,
+                    env=env,  # Pass the environment with DJANGO_SETTINGS_MODULE
+                    check=True,
+                    stdout=report_file,
+                    stderr=subprocess.STDOUT,
+                    text=True
+                )
+
+            # Extract results
+            uncovered_model_count, pii_check_passed_env, full_log = _extract_missing_pii_annotations(run_output_file)
+            env_report.append((
+                uncovered_model_count,
+                full_log,
+            ))
+
+        except BuildFailure as error_message:
+            fail_quality(pii_report_name, f'FAILURE: {error_message}')
+
+        # Update pii_check_passed based on the result of the current environment
+        if not pii_check_passed_env:
+            pii_check_passed = False
+
+    # If the PII check failed in any environment, fail the task
+    if not pii_check_passed:
+        fail_quality('pii', full_log)
+    else:
+        print("Successfully ran pii_check")
```

Comment:
This function was giving a false-positive. That is, the check was passing even when there were PII errors. This is because the `code_annotations` shell command is missing the arguments `--lint`- and `--coverage`, which tell it to raise an error if PII linting or coverage fails.

Unfortunately, the exiting Paver PII check _also_ raises false-positives! Both the Paver code and your new code write annotation checks to a file, and the use regex to read the result out of that file and determine whether to pass the check. However, if the report-generation command itself fails, then the check will pass. This has been happening for several years, during which time we've been building up a pile of PII annotation violations.

So:

* We have PII violations on master. Brian Mesick will be looking into the possibility of fixing some of those. For this PR though, please lower the [PII annotation threshold](https://github.com/openedx/edx-platform/blob/master/.pii_annotations.yml#L4) down from 94.5 to 71.6 so that it can pass on the existing codebase. You will also need to remove `celery_utils.FailedTask` from the `.annotation_safe_list`. When you are done with that, you should be able to run `make pii_check` locally and see that it passes.
* Whenever possible, we really ought to stop this dance of writing reports to files and then parsing the results with regex. It is complicated and mistake-prone. Instead, we should _just run the command_, let its output flow out to the user, and then use only the command's _exit code_ (via CalledProcessError) in order to determine whether to fail the CI check or not. Below is my suggestion of how to do that for the pii_check function.
```suggestion
def run_pii_check():
    """
    Guarantee that all Django models are PII-annotated.
    """
    failed = False
    for env_name, env_settings_file in (("CMS", "cms.envs.test"), ("LMS", "lms.envs.test")):
        try:
            print()
            print()
            print(f"Running {env_name} PII Annotation check and report")
            print("-" * 45)
            command = [
                "code_annotations",
                "django_find_annotations",
                "--config_file", ".pii_annotations.yml",
                "--app_name", "lms",
                "--lint",
                "--coverage",
            ]
            print(shlex.join(command))
            subprocess.run(
                command,
                env={
                    **os.environ,  # Include the current environment variables
                    "DJANGO_SETTINGS_MODULE": env_settings_file  # Set DJANGO_SETTINGS_MODULE for each environment
                },
                check=True,
                text=True,
            )
        except subprocess.CalledProcessError:
            # Error details should be printed to stdout, no need to capture them here.
            failed = True
    print()
    if failed:
        fail_quality("pii_check", 'PII annotation check failed. See details above.')
    else:
        print("Successfully checked PII annotations.")
```